### PR TITLE
chapter_9

### PR DESCRIPTION
**Pull request by chuan97sun**

### CODE REVIEW

**Comment by chuan97sun** [comment]
READY

File: app/controllers/sessions_controller.rb [hunk 1]
```diff
@@ -5,6 +5,7 @@ def create
     user = User.find_by email: params[:session][:email].downcase
     if user&.authenticate(params[:session][:password])
       log_in user
+      params[:session][:remember_me] == '1' ? remember(user) : forget(user)
```
Comment (saobv-2223) [inline app/controllers/sessions_controller.rb]: Please, move 1 to settings file
using .eql? instead of ==

File: app/models/user.rb [hunk 2]
```diff
@@ -17,17 +19,35 @@ class User < ApplicationRecord
...
     end
   end
+
+  def User.new_token
+    SecureRandom.urlsafe_base64
+  end
+
+  def remember
+   self.remember_token = User.new_token
+   update_attribute(:remember_digest, User.digest(remember_token))
```
Comment (saobv-2223) [inline app/models/user.rb]: remove ()

File: app/models/user.rb [hunk 3]
```diff
@@ -17,17 +19,35 @@ class User < ApplicationRecord
...
+  end
+
+  def forget
+    update_attribute :remember_digest, nil
+  end
+
+  def authenticated?(remember_token)
+    return false if remember_digest.nil?
+    BCrypt::Password.new(remember_digest).is_password?(remember_token)
```
Comment (saobv-2223) [inline app/models/user.rb]: remove ()

**Comment by chuan97sun** [comment]
ready

File: app/controllers/sessions_controller.rb [hunk 4]
```diff
@@ -3,8 +3,9 @@ def new; end
 
   def create
     user = User.find_by email: params[:session][:email].downcase
-    if user&.authenticate(params[:session][:password])
+    if user&.authenticate params[:session][:password]
       log_in user
+      params[:session][:remember_me] .eql? Settings.source.collection.session_controller_create ? remember(user) : forget(user)
```
Comment (manhnt-0960) [inline app/controllers/sessions_controller.rb]:
```suggestion
      params[:session][:remember_me].eql? Settings.source.collection.session_controller_create ? remember(user) : forget(user)
```

File: app/models/user.rb [hunk 5]
```diff
@@ -17,17 +19,35 @@ class User < ApplicationRecord
...
+    def digest string
       cost = ActiveModel::SecurePassword.min_cost ? BCrypt::Engine::MIN_COST :
         BCrypt::Engine.cost
       BCrypt::Password.create(string, cost: cost)
     end
   end
+
+  def User.new_token
+    SecureRandom.urlsafe_base64
+  end
```
Comment (manhnt-0960) [inline app/models/user.rb]: move this method in class<<self

File: app/models/user.rb [hunk 6]
```diff
@@ -17,17 +19,35 @@ class User < ApplicationRecord
...
+   self.remember_token = User.new_token
+   update_attribute :remember_digest, User.digest(remember_token)
+  end
+
+  def forget
+    update_attribute :remember_digest, nil
+  end
+
+  def authenticated? remember_token
+    return false if remember_digest.nil?
```
Comment (manhnt-0960) [inline app/models/user.rb]:
```suggestion
    return false unless remember_digest
```

**Comment by ngocvt-0484** [comment]
commented

**Comment by chuan97sun** [comment]
ready

**Comment by chuan97sun** [comment]
ready test

**Comment by chuan97sun** [comment]
ready

**Comment by ngocvt-0484** [comment]
merged please

**Comment by vanvtt-0952** [comment]
merged

**Pull request merged**